### ISSUE TITLE
[Bug]: Sqlite parse error / unexpected token:

### ISSUE THREAD

**Issue by ClundXIII** [sqlitebrowser/sqlitebrowser#3096]
### What did you do?

I opened a sqlite3 database, in the database structure overview the table structure is shown as empty. The Database was initialized by django 4.

Version information:
```
Version 3.11.2
Erstellt für x86_64-little_endian-lp64, running on x86_64
Qt-Version 5.12.5
SQLite-Version 3.31.1
```

### What did you expect to see?

According to sqlite console output the structure is the following:

```sql
CREATE TABLE IF NOT EXISTS "a_b" (
    "id" integer NOT NULL PRIMARY KEY,
    "name" text NOT NULL,
    "description" text NOT NULL,
    "field_json" text NOT NULL CHECK ((JSON_VALID("field_json") OR "field_json" IS NULL)),
    "field_integer" integer NOT NULL,
    "field_float" real NOT NULL,
    "foreign_id_id" integer NOT NULL REFERENCES "a_c" ("id") DEFERRABLE INITIALLY DEFERRED);
```

(I renamed the fields and added formatting)

### What did you see instead?

an empty table structure, I cannot "unfold" the table, no arrow next to it.

### DB4S Version

3.11.x

### What OS are you seeing the problem on?

Linux

### OS version

Linux Mint 20.3 Una, based on Ubuntu 20.04

### Relevant log output

```shell
Sqlite parse error: line 1:187: unexpected token: OR(CREATE TABLE "a_b" ("id" integer NOT NULL PRIMARY KEY, "name" text NOT NULL, "description" text NOT NULL, "field_json" text NOT NULL CHECK ((JSON_VALID("field_json") OR "field_json" IS NULL)), "field_integer" integer NOT NULL, "field_float" real NOT NULL, "foreign_id_id" integer NOT NULL REFERENCES "a_c" ("id") DEFERRABLE INITIALLY DEFERRED))
```

### Prevention against duplicate issues

- [X] I have searched for similar issues

**Comment by scottfurry** [comment]
Although SQLite is robust and quick, it does have some caveats and quirks that will catch the unaware. From [Wikipedia](https://en.wikipedia.org/wiki/SQLite#Features), SQLite implements the SQL-92 standard. This is fine for most use cases. However, some bleeding edge commands or details may not be implemented within SQLite. I have encountered this condition when moving data between SQLite and MySQL/PostgreSQL.

Each RDBMS or SQL implementation also has its own unique add-ons that are not apart of the SQL standard. Knowing about implementation-specific quirks for the different systems will save you on heartache and grief down the road.

A quick search indicated that there are settings/tweaks for [SQLite within Django](https://docs.djangoproject.com/en/4.0/ref/databases/#sqlite-notes). I recommend double checking settings to ensure export is SQLite compliant and not some other DB system.

tl;dr - there's SQL standard and then there is the DB software's idea of standard. The SQL standard tends to be the floor.

I suspect that the compound OR statement is causing SQLite grief. Since SQLite barfs on parsing the database structure, the failure to read the database progresses upwards to DB4S.

I suggest that you attempt to open the database file from the command line first. See [here](https://www.sqlitetutorial.net/sqlite-commands/) for help with command line usage. If the sqlite3 program chokes on the attempt then I believe you have your answer. There's a problem with the export of an SQL database from Django to SQLite.

**Comment by ClundXIII** [comment]
> I suggest that you attempt to open the database file from the command line first.

I did, that is how I got the underlying database structure.

**Comment by ClundXIII** [comment]
I think it is the `JSON_VALID`, which requires the `json1` extension for sqlite. I suspect this is not yet implemented in sqlitebrowser.

**Comment by FriedrichFroebel** [comment]
According to the initial report, the SQLite version is 3.31.1. The version seems to be from the official repositories at https://packages.ubuntu.com/focal/sqlitebrowser. Looking at the related SQLite library (https://packages.ubuntu.com/focal/libsqlite3-0), the source package file `rules` seems to indicate that JSON support should be available due to the *CLFAGS* including `-DSQLITE_ENABLE_JSON1`. I am not completely sure, but the fact that loading has been successful from the SQLite CLI seems to indicate that this indeed is valid.

As already mentioned, this most likely is an incomplete grammar/implementation for our lexer/parser.

**Comment by MKleusberg** [comment]
As the others already said, the error you get is a parser error. It doesn't recognise the ```(JSON_VALID("field_json") OR "field_json" IS NULL)``` bit. For this sort of problem it doesn't even matter whether the json extension is enabled or not because the parser error happens before any of this is executed.

However, in version 3.12.0 we have rewritten the entire parser and since then it is way more robust. As a matter of fact I have just tested your table with the latest version of DB4S and it works just fine. Have you any chance to update to 3.12.x on Mint? You can also just run the AppImage from [here](https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous). Does that work for you? :smile:

**Comment by ClundXIII** [comment]
> Have you any chance to update to 3.12.x on Mint? You can also just run the AppImage from [here](https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous). Does that work for you? smile

Yep, that version with Appimage works. And the Mint Upgrade to ubuntu 22.04 will drop in a few days anyway so I won't even need to use Appimage for that long :)
